numpy/core/src/multiarray/mapping.h

```diff
@@ -69,5 +84,57 @@ PyArray_MapIterNew(npy_index_info *indices , int index_num, int index_type,
                    PyArrayObject *arr, PyArrayObject *subspace,
                    npy_uint32 subspace_iter_flags, npy_uint32 subspace_flags,
                    npy_uint32 extra_op_flags, PyArrayObject *extra_op,
-                   PyArray_Descr *extra_op_dtype);
+                   PyArray_Descr *extra_op_dtype, int outer_indexing);
+
+/*
+ * Prototypes for attribute indexing helper (arr.oindex, etc.)
+ */
+typedef struct {
+        PyObject_HEAD
+        /*
+         * Attribute information portion.
+         */
+        PyArrayObject *array;
+        int indexing_method;  /* See mapping.h */
+} PyArrayAttributeIndexer;
+
+
+NPY_NO_EXPORT PyObject *
+PyArray_AttributeIndexerNew(PyArrayObject *array, int indexing_method);
+
+
+/*
+ * Prototypes for multi index objects passed to subclasses by arr.oindex, etc.
+ */
+typedef struct {
+        PyObject_HEAD
+        /*
+         * Attribute information portion.
+         */
+        PyObject *index;            /* The indexing object */
+        int indexing_method;        /* See mapping.h */
+        /* If bound is 1, the following are information about the array */
+        int bound;
```

Comment:
What purpose does binding serve?

Reply:
You do not have to pass in the shape and dtype again if you are a subclass, but I guess forcing the subclass to write `prepared(self.shape, self.dtype)` or `.prepared(self)` is OK too.

Reply:
Whose job it is to do the binding seems kinda orthogonal - all I'm saying is that this would be better represented with multiple classes, rather than a single class with a bunch of flags - especially given the precedent for bound objects in python

Reply:
Fair point, its much like set and frozenset. Probably should only have the bound version anyway, and expose all of these as attributes and just allow to create a new one from an old one with a new shape (and dtype), could be a method or not.

Got a bit lost in the fastest way to return something instead of thinking about how is nicest. My only real interest is currently if such info can help subclasses or not.